sql: panic on rollbackSQLTransaction with non-open txn

I got the following crash when killing a load generator app.

```
panic: rollbackSQLTransaction called on non-open txn: &{db:{sender:0xc8202dc160 userPriority:1 txnRetryOptions:{InitialBackoff:50000000 MaxBackoff:5000000000 Multiplier:2 MaxRetries:0 RandomizationFactor:0 Closer:<nil>}} wrapped:0xc820378e40 Proto:"sql/executor.go:380 sql txn" id=42e8c764 key=/Table/51/1/1/2/1 rw=true pri=0.00170982 iso=SERIALIZABLE stat=PENDING epo=1 ts=1459375813.964082583,1 orig=1459375813.964082583,1 max=1459375814.207097056,0 wto=false UserPriority:NORMAL Context:context.Background.WithCancel CollectedSpans:[] systemConfigTrigger:false retrying:false finalized:false}

goroutine 6503 [running]:
panic(0x4edbfa0, 0xc8206edf10)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql.rollbackSQLTransaction(0xc8207fc828, 0xc8207fc878, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:813 +0x1cf
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInAbortedTxn(0xc8202241c0, 0x7acf2c0, 0x5dd3b78, 0xc8207fc828, 0xc8207fc878, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:603 +0x166
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8202241c0, 0xc8206edc40, 0x1, 0x1, 0xc8207fc878, 0xc8207fc828, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:558 +0xbbb
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8202241c0, 0xc8207fc878, 0x3, 0xc8207fc828, 0xc8207497a0, 0xc8206edc40, 0x1, 0x1, 0xc820a093a0, 0xc820a093b8, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:488 +0x147
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc8202dc160, 0xc8207497a0, 0xc8208b9c80)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:401 +0x325
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc8202dc160, 0x55d0000, 0x0, 0x0, 0xc820a094f8, 0x0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/client/txn.go:531 +0x410
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8202241c0, 0x7a059d0, 0xc820644340, 0xc8207fc800, 0xc8207be137, 0x8, 0x0, 0x0, 0x0, 0x7a50d00)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:405 +0x892
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8202241c0, 0x7a059d0, 0xc820644340, 0xc8207fc800, 0xc8207be137, 0x8, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/executor.go:295 +0xe2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc8202f4600, 0x7a059d0, 0xc820644340, 0xc8207be137, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:604 +0xb2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc8202f4600, 0x7a059d0, 0xc820644340, 0xc8202f4628, 0x0, 0x0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:314 +0xf2
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8202f4600, 0x0, 0x0, 0x0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:270 +0x10d3
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc8202c7e58, 0x7a083d8, 0xc820241810, 0x0, 0x0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:159 +0x929
github.com/cockroachdb/cockroach/server.(*Server).Start.func4.1(0x7a083d8, 0xc820241810)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/server/server.go:300 +0x4b
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc820140600, 0x7a083d8, 0xc820241810, 0xc820200020)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/util/net.go:127 +0x31f
```

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -809,8 +809,9 @@ func rollbackSQLTransaction(txnState *txnState, p *planner) Result {
 	if p.txn != txnState.txn {
 		panic("rollbackSQLTransaction called on a different txn than the planner's")
 	}
-	if txnState.State != Open {
-		panic(fmt.Sprintf("rollbackSQLTransaction called on non-open txn: %+v", txnState.txn))
+	if txnState.State != Open && txnState.State != RestartWait {
+		panic(fmt.Sprintf("rollbackSQLTransaction called on txn in wrong state: %s (txn: %s)",
+			txnState.State, txnState.txn.Proto))
 	}
 	pErr := p.txn.Rollback()
 	result := Result{PGTag: (*parser.RollbackTransaction)(nil).StatementTag()}

--- a/sql/pgwire/v3.go
+++ b/sql/pgwire/v3.go
@@ -213,13 +213,12 @@ func (c *v3Conn) serve(authenticationHook func(string, bool) error) error {
 			c.writeBuf.initMsg(serverMsgReady)
 			var txnStatus byte
 			switch c.session.TxnState.State {
-			case sql.Aborted:
+			case sql.Aborted, sql.RestartWait:
+				// We send status "InFailedTransaction" also for state RestartWait
+				// because GO's lib/pq freaks out if we invent a new status.
 				txnStatus = 'E'
 			case sql.Open:
 				txnStatus = 'T'
-			case sql.RestartWait:
-				// This state is not part of the Postgres protocol.
-				txnStatus = 'R'
 			case sql.NoTxn:
 				// We're not in a txn (i.e. the last txn was committed).
 				txnStatus = 'I'

--- a/sql/txn_restart_test.go
+++ b/sql/txn_restart_test.go
@@ -627,3 +627,51 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 		t.Fatal(err)
 	}
 }
+
+// TestRollbackInRestartWait ensures that a ROLLBACK while the txn is in the
+// RetryWait state works.
+func TestRollbackInRestartWait(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx, cmdFilters := createTestServerContext()
+	server, sqlDB, _ := setupWithContext(t, ctx)
+	defer cleanup(server, sqlDB)
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (k TEXT PRIMARY KEY, v TEXT);
+`); err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up error injection that causes retries.
+	magicVals := createFilterVals(nil, nil)
+	magicVals.endTxnRestartCounts = map[string]int{
+		"boulanger": 1,
+	}
+	cmdFilters.AppendFilter(
+		func(args storageutils.FilterArgs) *roachpb.Error {
+			if err := injectErrors(args.Req, args.Hdr, magicVals); err != nil {
+				return roachpb.NewErrorWithTxn(err, args.Hdr.Txn)
+			}
+			return nil
+		}, false)
+
+	tx, err := sqlDB.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tx.Exec("SAVEPOINT cockroach_restart"); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tx.Exec(
+		"INSERT INTO t.test (k, v) VALUES ('g', 'boulanger')"); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tx.Exec("RELEASE SAVEPOINT cockroach_restart"); err == nil {
+		t.Fatal("expected RELEASE to fail")
+	}
+	if err := tx.Rollback(); err != nil {
+		t.Fatal(err)
+	}
+}

